DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp

```diff
@@ -0,0 +1,1852 @@
+/////////////////////////////////
+
+// Object stores oneAPI shim metrics
+// Member Functions (details closer to function definition):
+// ShimMetrics - Constructor; initializes all metrics and obtains maximum device
+// allocation and maxmum device global memory TestGlobalMem - Host to device
+// global memory interface check HostSpeed - Host to device global memory
+// bandwidth measurement HostRWTest - Unaligned read & writes from host to
+// device global memory KernelClkFreq - Kernel clock frequency measurement
+// KernelLatency - Kernel latency measurement
+// KernelLaunchTest - Host to kernel interface check
+// KernelMemRW - Kernel to device global memory interface check
+// KernelMemBW - Kernel to device global memory bandwidth measurement
+
+class ShimMetrics {
+ public:
+  ShimMetrics(queue &q)
+      : h2d_rd_bw{0},
+        h2d_wr_bw{0},
+        h2d_rd_wr_bw{0},
+        h2d_rw_test{false},
+        kernel_freq{0},
+        kernel_latency{0},
+        kernel_thruput{0},
+        kernel_mem_bw{0},
+        kernel_mem_rw_test{false} {
+    max_buffer_size = q.get_device().get_info<info::device::global_mem_size>();
+    max_alloc_size =
+        q.get_device().get_info<info::device::max_mem_alloc_size>();
+    std::cout << "\nclGetDeviceInfo CL_DEVICE_GLOBAL_MEM_SIZE = "
+              << max_buffer_size << "\n";
+    std::cout << "clGetDeviceInfo CL_DEVICE_MAX_MEM_ALLOC_SIZE = "
+              << max_alloc_size << "\n";
+    std::cout << "Device buffer size available for allocation = "
+              << max_alloc_size << " bytes\n";
+  }
+
+  ~ShimMetrics() {}
+
+  size_t TestGlobalMem(queue &q);
+  int HostSpeed(queue &q);
+  int HostRWTest(queue &q, size_t dev_offset = 0);
+  int KernelClkFreq(queue &q, bool report_chk = true);
+  int KernelLatency(queue &q);
+  int KernelLaunchTest(queue &q);
+  int KernelMemRW(queue &q);
+  int KernelMemBW(queue &q);
+
+ private:
+  float h2d_rd_bw;
+  float h2d_wr_bw;
+  float h2d_rd_wr_bw;
+  bool h2d_rw_test;
+  float kernel_freq;
+  float kernel_latency;
+  float kernel_thruput;
+  float kernel_mem_bw;
+  bool kernel_mem_rw_test;
+  cl_ulong max_buffer_size;
+  cl_ulong max_alloc_size;
+};
+
+//////////////////////////////////////
+// **** TestGlobalMem function **** //
+//////////////////////////////////////
+
+// Input:
+// queue &q - queue to submit operation
+// Returns:
+// Number of errors in transfer OR 1 if device memory allocation is 0 or test
+// fails (return 0 means test passed)
+
+// The function does the following tasks:
+// 1. Get maximum device global memory allocation size
+// 2. Allocate host memory to use to store data to be written to and read from
+// device
+// 3. Write to device global memory
+// 4. Read from device global memory
+// 5. Verify data read back matches value written
+// 6. Report read, write bandwidth
+// If this test passes (returns 0), the host to device global memory interface
+// is working fine
+
+size_t ShimMetrics::TestGlobalMem(queue &q) {
```

Comment:
why are the function implementations in this `.hpp` file? I thought normally implemetnations went in `.cpp` files.

Reply:
Normally. But that would pollute the CPP file. And building multiple CPP files is annoying. So... take your pick :)